### PR TITLE
Features: Conditional Offer Constraints, Custom Env Variables, Node Taints are applied from K3s Arguments

### PR DESCRIPTION
**Pull request by itsoksarvesh**
Feature Implementation

- Conditional Hostname Offer Constraints... Will work only if Environment variable for conditional offer constraint is set and if Hostname constraints are set for all of DS, Server and Agent.. 
- Custom Env Variables - Feature to add custom env variables to the K3s Nodes while starting. Useful for having dynamic values per framework

Bugfix

- Delayed Node Taints Application - Node Taints are now applied from K3s Arguments. Older way of adding Node taints from controller is not yet decommissioned but will be decommisioned in the future unless there is a requirement for that. Currently the older way acts as a backup solution
- Offer Accepted even when required ports are not available
- Added Kafka Plugin Call for Mesos Subscribe Event

### CODE REVIEW

**Review by andreaspeters — CHANGES_REQUESTED** [review]

File: init.go [hunk 1]
```diff
@@ -219,3 +244,16 @@ func loadPlugins(r *redis.Redis) {
 func stringToBool(par string) bool {
 	return strings.Compare(par, "true") == 0
 }
+
+func getUniqueStringList(slice []string) []string {
```
[the remark this answers is not in the document]
Comment (itsoksarvesh) [reply init.go:242]: Added a comment explaining the use of the function

File: scheduler/prepare.go [hunk 2]
```diff
@@ -154,6 +154,11 @@ func (e *Scheduler) prepareTaskInfoExecuteContainer(agent *mesosproto.AgentID, c
 		msg.HealthCheck = cmd.Health
 	}
 
+	// msg.Limits = map[string]*mesosproto.Value_Scalar{
+	// 	"cpus": {Value: &cmd.CPU},
+	// 	"mem":  {Value: &cmd.Memory},
+	// }
+
```
[the remark this answers is not in the document]
Comment (itsoksarvesh) [reply scheduler/prepare.go]: Done

File: app.go [hunk 3]
```diff
@@ -64,6 +64,7 @@ func main() {
 
 	framework.FrameworkInfo.User = util.StringToPointer(framework.FrameworkUser)
 	framework.FrameworkInfo.Name = util.StringToPointer(framework.FrameworkName)
+	framework.FrameworkInfo.Role = &framework.FrameworkRole
```
[the remark this answers is not in the document]
Comment (itsoksarvesh) [reply app.go]: Changed to FrameworkInfo.Roles.. Needed to add MultiRole Capability for Roles & Offer constraints to work

File: scheduler/scheduler.go [hunk 4]
```diff
@@ -267,7 +306,7 @@ func (e *Scheduler) reconcile() {
 				Value: &task.MesosAgent.ID,
 			},
 		})
-		logrus.WithField("func", "mesos.Reconcile").Debug("Reconcile Task: ", task.TaskID)
+		logrus.WithField("func", "mesos.Reconcile").Debug("Reconcile Task: ", task.TaskID, task.TaskName)
```
[the remark this answers is not in the document]
Comment (itsoksarvesh) [reply scheduler/scheduler.go]: Changed to Debugf

File: init.go [hunk 5]
```diff
@@ -98,6 +98,8 @@ func init() {
 	config.DSMaxRestore = 0
 	config.K3SAgentMaxRestore = 0
 	config.K3SServerMaxRestore = 0
+	config.EnableHostnameOfferConstraint = stringToBool(util.Getenv("ENABLE_HOSTNAME_OFFER_CONSTRAINT", "true"))
```
[the remark this answers is not in the document]
Comment (itsoksarvesh) [reply init.go]: Agreed.  Made the necessary changes

File: init.go [hunk 6]
```diff
@@ -177,6 +179,29 @@ func init() {
 	} else {
 		config.PluginsEnable = true
 	}
+
+	// Hostname Offer constraints
+	if config.EnableHostnameOfferConstraint {
+		if config.K3SServerConstraintHostname == "" || config.K3SAgentConstraintHostname == "" || config.DSConstraintHostname == "" {
```
[the remark this answers is not in the document]
Comment (itsoksarvesh) [reply init.go]: Agreed.  Made the necessary changes

**Comment by itsoksarvesh** [comment]
@andreaspeters My latest 2 commits on the iss-lab/mesos-m3s repo have also been added to this PR automatically since it was not closed earlier... The commits are for the following features

- Setting Mesos Resource allocation Limits on the M3s Tasks ... There are individual ENV variables that control the limits as well as a global flag which will control the enforcement of the limits .. I believe both are needed ... The global flag gives us control to set/unset limits whereas the individual env variables control what the limit is set for allowing for over-provisioning etc.
- Bugfix in the M3s Status API if kubernetes is not up.. There was a null pointer panic from kubernetes client...

**Pull request merged**